vlog4j-core/src/main/java/org/semanticweb/vlog4j/core/reasoner/implementation/VLogReasoner.java

```diff
@@ -808,4 +811,34 @@ ReasonerState getReasonerState() {
 	void setReasonerState(ReasonerState reasonerState) {
 		this.reasonerState = reasonerState;
 	}
+
+	@Override
+	public void getKbFacts(OutputStream stream) throws IOException {
+		// TODO Auto-generated method stub
+		HashSet<PositiveLiteral> headLiterals = new HashSet<PositiveLiteral>();
+		for (Rule rule : this.knowledgeBase.getRules()) {
+			for (PositiveLiteral positiveLiteral : rule.getHead()) {
+				headLiterals.add(positiveLiteral);
+			}
+		}
+		for (PositiveLiteral positiveliteral : headLiterals) {

```

Comment:
you may write the same fact several times: consider rules
```
a(x) ->b(x,x)
b(x,y)->b(y,x)
```
Or even
```
a(x) -> b(x)
a(y) -> b(y)
```
Where b(x) and b(y) are different PositiveLiterals.